### PR TITLE
Add memory pool kind and improve the management in memory manager

### PR DESCRIPTION
**Pull request by xiaoxmeng**
- Add memory kind support
    -   kLeaf is used for memory allocation. User can allocate memory from this
         kind of pool but can't create child pool from it.
    -   kAggregation is used to manage the memory pool hierarchy and aggregate
         the memory usage from the leaf pools. The user can't directly allocate memory
         from this kind of pool but can create child pools from it.
- Add a flag in MemoryUsageTracker to indicate if it is associated with a leaf
   memory pool or not. Only allow memory reservation leaf memory tracker
- Improve the memory pool management in memory manager to fit the memory
   arbitration needs and for later merge of MemoryPool and MemoryUsageTracker:
   -  Instead of having a root memory pool in the memory manager and create all
        the user requested memory pool as a child from the root pool, we create each
       user requested aggregation memory pool (which is the actual root memory pool
       or the user) as a root memory pool and keeps track of its raw reference inside
       the memory manager. The user owns the aggregation memory pool and the
       memory manager sets a destruction callback to remove the reference track when
       the pool destroys.
   -  Remove the getRoot() to access the root memory pool inside the memory manager.
       The root memory pool is now used as the default memory pool management. The
       user can create a leaf memory pool through the new getPool which creates a child
       pool from the internal root memory pool. For use case such as spilling system
       memory pool and tests, and some early Velox memory integrations in Meta internally.
       The user owns the created leaf memory pool.
   -  Provides a deprecatedGetPool() to get the default leaf memory pool which is created
      as the child of the default root memory pool on memory manager construction. The
      memory manager owns the pool. This will be deprecated and is for Meta internally
      legacy use cases.

### CODE REVIEW

File: velox/buffer/tests/BufferTest.cpp [hunk 1]
```diff
@@ -41,7 +41,8 @@ static_assert(!Buffer::is_pod_like_v<std::shared_ptr<int>>, "");
 class BufferTest : public testing::Test {
  protected:
   void SetUp() override {
-    pool_ = memoryManager_.getChild();
+    pool_ =
+        memoryManager_.getPool("BufferTest", memory::MemoryPool::Kind::kLeaf);
```
Comment (mbasmanova) [inline velox/buffer/tests/BufferTest.cpp:45]: nit: Would it make sense to add getLeafPool method?

```
memoryManager_.getLeafPool("BufferTest");
```

File: velox/common/memory/Memory.cpp [hunk 2]
```diff
@@ -60,13 +68,51 @@ uint16_t MemoryManager::alignment() const {
   return alignment_;
 }
 
-MemoryPool& MemoryManager::getRoot() const {
-  return *root_;
+std::shared_ptr<MemoryPool> MemoryManager::getPool(
+    const std::string& name,
+    MemoryPool::Kind kind,
+    int64_t cap) {
```
Comment (mbasmanova) [inline velox/common/memory/Memory.cpp:74]: nit: perhaps, use a more descriptive name for this parameter; maybe, maxBytes

File: velox/common/memory/Memory.h [hunk 3]
```diff
@@ -85,15 +85,23 @@ class IMemoryManager {
   /// Returns the memory allocation alignment of this memory manager.
   virtual uint16_t alignment() const = 0;
 
-  /// Power users that want to explicitly modify the tree should get the root of
-  /// the tree.
+  /// Creates a memory pool for use with specified 'name', 'kind' and 'cap'. If
+  /// 'name' is missing, the memory manager generates a default name internally
```
Comment (mbasmanova) [inline velox/common/memory/Memory.h:89]: > generates a default name internally to ensure uniqueness

If 'name' is not empty, is the caller expected to ensure uniqueness?

File: velox/common/memory/Memory.h [hunk 4]
```diff
@@ -140,32 +148,56 @@ class MemoryManager final : public IMemoryManager {
...
 
   bool reserve(int64_t size) final;
   void release(int64_t size) final;
 
+  /// Returns the number of alive user memory pools including both root and leaf
+  /// memory pools.
+  ///
+  /// NOTE: this doesn't count the memory manager's internal default root and
+  /// leaf memory pools.
+  size_t numUserPools() const;
```
Comment (mbasmanova) [inline velox/common/memory/Memory.h:168]: Any particular reason to use 'User' in the name of this method. I'd expect numPools() name to match getPool().

File: velox/common/memory/Memory.cpp [hunk 5]
```diff
@@ -60,13 +68,51 @@ uint16_t MemoryManager::alignment() const {
...
+  options.alignment = alignment_;
+  options.capacity = cap;
+  auto pool = std::make_shared<MemoryPoolImpl>(
+      this,
+      poolName,
+      MemoryPool::Kind::kAggregate,
+      nullptr,
+      poolDestructionCb_,
+      options);
+  folly::SharedMutex::WriteHolder guard{mutex_};
```
Comment (mbasmanova) [inline velox/common/memory/Memory.cpp:95]: Why not `std::lock_guard<std::mutex>` ?

**Review by mbasmanova — COMMENTED** [review]
@xiaoxmeng I haven't finished reading this PR as I ran out of time. Flushing comments accumulated so far.

File: velox/buffer/tests/BufferTest.cpp (hunk 1 above)
Comment (xiaoxmeng) [reply velox/buffer/tests/BufferTest.cpp:45]: As discussed offline, we will leave API naming changes in followup.

File: velox/common/memory/Memory.cpp (hunk 5 above)
Comment (xiaoxmeng) [reply velox/common/memory/Memory.cpp:95]: Use r/w lock here for memory arbitration support. Each memory arbitration request holds a reader lock to allow concurrent access to the memory arbitration (the memory arbitrartor can serialize the arbitration requests internally) and the thread safety of the memory pools_. Also user can access memory manager like check the number of pools without blocking.

File: velox/common/memory/Memory.h (hunk 4 above)
Comment (xiaoxmeng) [reply velox/common/memory/Memory.h:168]: The default memory pool is not counted. Yeah, numPools are better. Changed.

File: velox/common/memory/Memory.cpp [hunk 6]
```diff
@@ -27,24 +27,32 @@ namespace {
...
+      deprecatedDefaultLeafPool_(defaultRoot_->addChild(
+          kDefaultLeafName.str(),
+          MemoryPool::Kind::kLeaf)) {
   VELOX_CHECK_NOT_NULL(allocator_);
   VELOX_USER_CHECK_GE(memoryQuota_, 0);
   MemoryAllocator::alignmentCheck(0, alignment_);
 }
 
 MemoryManager::~MemoryManager() {
+  VELOX_CHECK_EQ(numUserPools(), 0);
```
[the remark this answers is not in the document]
Comment (xiaoxmeng) [reply velox/common/memory/Memory.cpp:55]: Added.

File: velox/common/memory/Memory.h (hunk 3 above)
Comment (xiaoxmeng) [reply velox/common/memory/Memory.h:89]: Changed to capacity?

File: velox/common/memory/Memory.cpp (hunk 2 above)
Comment (xiaoxmeng) [reply velox/common/memory/Memory.cpp:74]: Changed.

File: velox/common/memory/Memory.cpp [hunk 7]
```diff
@@ -60,13 +68,51 @@ uint16_t MemoryManager::alignment() const {
...
+  folly::SharedMutex::WriteHolder guard{mutex_};
+  auto it = userPools_.begin();
+  while (it != userPools_.end()) {
+    if (*it == pool) {
+      userPools_.erase(it);
+      return;
+    }
+    ++it;
+  }
+  VELOX_UNREACHABLE("Memory pool{} is not found", pool->name());
```
Comment (mbasmanova) [inline velox/common/memory/Memory.cpp:111]: Is it save to call 'pool-name()' here? Can pool be a dangling pointer?

File: velox/common/memory/Memory.h (hunk 3 above)
Comment (xiaoxmeng) [reply velox/common/memory/Memory.h:89]: User might not care but we internally use atomic id to ensure the uniqueness for debugging purpose? Thanks!

File: velox/connectors/tpch/TpchConnector.cpp [hunk 8]
```diff
@@ -31,21 +31,21 @@ RowVectorPtr getTpchData(
     memory::MemoryPool* pool) {
   switch (table) {
     case Table::TBL_PART:
-      return velox::tpch::genTpchPart(maxRows, offset, scaleFactor, pool);
+      return velox::tpch::genTpchPart(pool, maxRows, offset, scaleFactor);
```
Comment (mbasmanova) [inline velox/connectors/tpch/TpchConnector.cpp:34]: Unrelated change?

File: velox/tpch/gen/TpchGen.h [hunk 9]
```diff
@@ -80,11 +80,10 @@ TypePtr resolveTpchColumn(Table table, const std::string& columnName);
 ///  o_comment: VARCHAR
 ///
 RowVectorPtr genTpchOrders(
+    memory::MemoryPool* pool,
     size_t maxRows = 10000,
     size_t offset = 0,
-    double scaleFactor = 1,
-    memory::MemoryPool* pool =
```
Comment (mbasmanova) [inline velox/tpch/gen/TpchGen.h:86]: Any particular reason to move 'pool' parameter to the front? It seems just removing the default value would be sufficient.

**Review by mbasmanova — APPROVED** [review]
@xiaoxmeng Looks good % some comments.

File: velox/connectors/tpch/TpchConnector.cpp (hunk 8 above)
Comment (xiaoxmeng) [reply velox/connectors/tpch/TpchConnector.cpp:34]: It is related because the memory pool has been explicitly assigned with this change. Otherwise we have to use the deprecated API. So make this cleanup.

File: velox/tpch/gen/TpchGen.h (hunk 9 above)
Comment (xiaoxmeng) [reply velox/tpch/gen/TpchGen.h:86]: We can't because all the preceding input parameters have defaults :)
